### PR TITLE
[MODCON-30] - Exception handling for Assign/unassign a users affiliations

### PR DESCRIPTION
**Pull request by azizbekxm**
## Purpose

https://issues.folio.org/browse/UIU-2852
## Pre-Merge Checklist:
Before merging this PR, please go through the following list and take appropriate actions.

- Does this PR meet or exceed the expected quality standards?
  - [ ] Code coverage on new code is 80% or greater
  - [ ] Duplications on new code is 3% or less
  - [ ] There are no major code smells or security issues
- Does this introduce breaking changes?
  - [ ] Were any API paths or methods changed, added or removed?
  - [ ] Were there any schema changes?
  - [ ] Did any of the interface versions change?
  - [ ] Were permissions changed, added or removed?
  - [ ] Are there new interface dependencies?
  - [ ] There are no breaking changes in this PR.

If there are breaking changes, please **STOP** and consider the following:

- What other modules will these changes impact?
- Do JIRAs exist to update the impacted modules?
  - [ ] If not, please create them
  - [ ] Do they contain the appropriate level of detail?  Which endpoints/schemas changed, etc.
  - [ ] Do they have all the appropriate links to blocked/related issues?
- Are the JIRAs under active development?
  - [ ] If not, contact the project's PO and make sure they're aware of the urgency.
- Do PRs exist for these changes?
  - [ ] If so, have they been approved?
- Did you modify code to call some additional endpoints?
  - [ ] If so, do you check that necessary module permission added in ModuleDescriptor-template.yaml file?

Ideally, all the PRs involved in breaking changes would be merged on the same day
to avoid breaking the folio-testing environment.
Communication is paramount if that is to be achieved,
especially as the number of inter-module and inter-team dependencies increase.

While it's helpful for reviewers to help identify potential problems,
ensuring that it's safe to merge is ultimately the responsibility of the PR assignee.

### CODE REVIEW

**Comment by azizbekxm** [comment]
Example for Missing permission exception case
```
{
  "errors": [
    {
      "message": "Access for user 'ss_admin' (b82b46b6-9a6e-46f0-b986-5c643d9ba036) requires permission: users.item.get",
      "type": "-1",
      "code": "PERMISSION_ERROR"
    }
  ]
}
```

**Comment by azizbekxm** [comment]
some temporarily changes https://github.com/folio-org/mod-consortia/pull/35/commits/794c6fcda8406da0b1becbd7a1150fc3233715ae

File: src/main/java/org/folio/consortia/controller/ErrorHandlingController.java [hunk 1]
```diff
@@ -35,13 +36,13 @@ public class ErrorHandlingController {
   @ResponseStatus(HttpStatus.NOT_FOUND)
   @ExceptionHandler(ResourceNotFoundException.class)
   public Errors handleNotFoundException(ResourceNotFoundException e) {
-    return createInternalError(e.getMessage(), NOT_FOUND_ERROR);
+    return createExternalError(e.getMessage(), NOT_FOUND_ERROR);
   }
```
Comment (azizbekxm) [inline src/main/java/org/folio/consortia/controller/ErrorHandlingController.java:40]: improved error type(-1). NOT FOUND, DUPLICATE_ERROR... -> client error 4xx.

File: src/main/java/org/folio/consortia/controller/ErrorHandlingController.java [hunk 2]
```diff
@@ -54,28 +55,39 @@ public Errors handleDataIntegrityViolationException(DataIntegrityViolationExcept
...
+    log.error("Handle consortium client exception", e);
+    return createPermissionError(e, PERMISSION_REQUIRED);
+  }
+
+  @ResponseStatus(HttpStatus.BAD_GATEWAY)
+  @ExceptionHandler(IllegalStateException.class)
+  public Errors handleIllegalStateException(IllegalStateException e) {
+    log.error("Handle illegal state exception", e);
+    return createInternalError(e.getMessage(), BAD_GATEWAY);
   }
```
Comment (azizbekxm) [inline src/main/java/org/folio/consortia/controller/ErrorHandlingController.java:91]: 5xx -> error type "-2" internal error etc BAD_GATEWAY

File: src/main/java/org/folio/consortia/service/impl/UserTenantServiceImpl.java [hunk 3]
```diff
@@ -248,8 +250,10 @@ private User getUser(UUID userId) {
     } catch (FeignException.NotFound e) {
       log.info("User with userId {} does not exist in schema, going to use new one", userId);
       return new User();
+    } catch (FeignException.Forbidden e) {
+      throw new ConsortiumClientException(e);
     } catch (FeignException e) {
```
Comment (azizbekxm) [inline src/main/java/org/folio/consortia/service/impl/UserTenantServiceImpl.java:255]: permission exception is formatted with additional method inside of Error Helper

File: src/test/java/org/folio/consortia/controller/TenantControllerTest.java [hunk 4]
```diff
@@ -41,10 +42,15 @@ class TenantControllerTest extends BaseTest {
   ConsortiumRepository consortiumRepository;
```
Comment (azizbekxm) [inline src/test/java/org/folio/consortia/controller/TenantControllerTest.java:42]: in TenantControllerTest, all testest changed order (first success cases and second fail cases) and add more test cases

File: src/test/java/org/folio/consortia/service/UserTenantServiceTest.java [hunk 5]
```diff
@@ -55,12 +36,34 @@
...
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.doNothing;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
+
 @SpringBootTest
 @EnableAutoConfiguration(exclude = BatchAutoConfiguration.class)
 @EntityScan(basePackageClasses = UserTenantEntity.class)
 class UserTenantServiceTest {
```
Comment (azizbekxm) [inline src/test/java/org/folio/consortia/service/UserTenantServiceTest.java:62]: in UserTenantService, changed tests order (first success, following error cases) and added more cases and improved tests

File: descriptors/ModuleDescriptor-template.json [hunk 6]
```diff
@@ -109,7 +109,9 @@
           "modulePermissions": [
             "users.item.get",
             "users.item.put",
-            "users.item.post"
+            "users.item.post",
+            "perms.users.get",
+            "configuration.entries.collection.get"
           ]
```
Comment (azizbekxm) [inline descriptors/ModuleDescriptor-template.json:115]: in order to not have a problem with unassigning/assigning affiliations.

File: src/main/java/org/folio/consortia/utils/ErrorHelper.java [hunk 7]
```diff
@@ -33,13 +34,23 @@ public static Errors createInternalError(String message, ErrorCode errorCode) {
...
+
+  private String extractPermissionFromErrorMessage(String e) {
+    return e.substring(e.lastIndexOf('[') + 1, e.lastIndexOf(']')); // extract the main part from the error message
   }
 
   public enum ErrorType {
-    INTERNAL("-1"),
-    FOLIO_EXTERNAL_OR_UNDEFINED("-2"),
-    EXTERNAL_OR_UNDEFINED("-3"),
+    EXTERNAL("-1"), // bad request or client error
```
Comment (azizbekxm) [inline src/main/java/org/folio/consortia/utils/ErrorHelper.java:51]: -1 error for client error. for example bad request or illegal argument exception
-2 error for internal error. for example error with db or not connect with other module

File: src/main/java/org/folio/consortia/utils/TenantContextUtils.java [hunk 8]
```diff
@@ -48,7 +48,8 @@ public static void runInFolioContext(FolioExecutionContext context, Runnable run
   private static FolioExecutionContext getContextFromKafkaHeaders(MessageHeaders headers,
                                                                   FolioModuleMetadata moduleMetadata) {
     Map<String, Collection<String>> map = new HashMap<>();
-    map.put(XOkapiHeaders.TENANT, getHeaderValue(headers, XOkapiHeaders.TENANT, null));
+    //todo temporary solution, separate story will be created
+    map.put(XOkapiHeaders.TENANT, List.of("mobius"));
```
Comment (SerhiiNosko) [inline src/main/java/org/folio/consortia/utils/TenantContextUtils.java:52]: I will create story to fix this in the next sprint

**Review by SerhiiNosko — APPROVED** [review]

File: src/main/java/org/folio/consortia/utils/TenantContextUtils.java (hunk 8 above)
Comment (SerhiiNosko) [reply src/main/java/org/folio/consortia/utils/TenantContextUtils.java:52]: Story is created: https://issues.folio.org/browse/MODCON-32

File: src/main/java/org/folio/consortia/service/impl/UserTenantServiceImpl.java [hunk 9]
```diff
@@ -126,9 +127,8 @@ public UserTenant save(UUID consortiumId, UserTenant userTenantDto) {
 
     Optional<UserTenantEntity> userTenant = userTenantRepository.findByUserIdAndIsPrimary(userTenantDto.getUserId(), IS_PRIMARY_TRUE);
     if (userTenant.isEmpty()) {
-      log.warn("Could not find user with id: {} in user_tenant table for tenant id: {}",
-        userTenantDto.getUserId(), userTenantDto.getTenantId());
-      throw new ResourceNotFoundException(USER_ID, String.valueOf(userTenantDto.getUserId()));
+      log.warn("Could not find user '{}' with primary affiliation in user_tenant table", userTenantDto.getUserId());
+      throw new ResourceNotFoundException(String.format(NOT_FOUND_PRIMARY_AFFILIATION_MSG,USER_ID, userTenantDto.getUserId()));
```
Comment (tatsiana-tarhonskaya) [inline src/main/java/org/folio/consortia/service/impl/UserTenantServiceImpl.java:131]: missed space

**Review by tatsiana-tarhonskaya — APPROVED** [review]

File: src/test/java/org/folio/consortia/service/UserTenantServiceTest.java [hunk 10]
```diff
@@ -55,13 +36,33 @@
...
+import static org.junit.jupiter.api.Assertions.assertDoesNotThrow;
+import static org.junit.jupiter.api.Assertions.assertEquals;
+import static org.junit.jupiter.api.Assertions.assertNull;
+import static org.junit.jupiter.api.Assertions.assertThrows;
+import static org.mockito.ArgumentMatchers.any;
+import static org.mockito.ArgumentMatchers.anyString;
+import static org.mockito.Mockito.doNothing;
+import static org.mockito.Mockito.times;
+import static org.mockito.Mockito.verify;
+import static org.mockito.Mockito.when;
```
Comment (SerhiiNosko) [inline src/test/java/org/folio/consortia/service/UserTenantServiceTest.java:56]: We need to use the same checkstyle for all team

File: docs/api-guide.md [hunk 11]
```diff
@@ -1,28 +1,45 @@
...
+## Tenants API
+
+### Endpoints
 
+| METHOD | URL                 | Permission                       | DESCRIPTION                                   |
+|--------|---------------------|----------------------------------|-----------------------------------------------|
+| GET    | /tenants            | consortia.tenants.collection.get | Return list of tenants with `limit`, `offset` |
+| POST   | /tenants            | consortia.tenants.item.post      | Return saved tenant                           |
+| PUT    | /tenants/{tenantId} | consortia.tenants.item.put       | Return updated tenant based on `consortiumId` |
+| DELETE | /tenants/{tenantId} | consortia.tenants.item.delete    | Deleted tenant based on `consortiumId`        |
```
Comment (tatsiana-tarhonskaya) [inline docs/api-guide.md:23]: Deleted->Delete

File: docs/api-guide.md [hunk 12]
```diff
@@ -1,28 +1,45 @@
...
+|--------|--------------------------------------------------------|---------------------------------------|------------------------------------------------------------------------------------------------------|
+| GET    | /consortia/{consortiumId}/user-tenants                 | consortia.user-tenants.collection.get | Return list of user tenant associations based on `userId`, `username`, `tenantId`, `limit`, `offset` |
+| GET    | /consortia/{consortiumId}/user-tenants/{associationId} | consortia.user-tenants.item.get       | Return user-tenant with provided `associationId`                                                     |
+| POST   | /consortia/{consortiumId}/user-tenants                 | consortia.user-tenants.item.post      | Return saved user-tenant association                                                                 |
+| DELETE | /consortia/{consortiumId}/user-tenants                 | consortia.user-tenants.item.delete    | Deleted user-tenant association based on `userId` and `tenantId`                                     |
```
Comment (tatsiana-tarhonskaya) [inline docs/api-guide.md:34]: Deleted->Delete

File: src/main/java/org/folio/consortia/service/impl/UserTenantServiceImpl.java (hunk 9 above)
Comment (azizbekxm) [reply src/main/java/org/folio/consortia/service/impl/UserTenantServiceImpl.java:131]: sorry, I didn't catch it. could you please explain a little bit

Comment (azizbekxm) [reply src/main/java/org/folio/consortia/service/impl/UserTenantServiceImpl.java]: fixed

File: docs/api-guide.md (hunk 11 above)
Comment (azizbekxm) [reply docs/api-guide.md]: Fixed, thanks

File: docs/api-guide.md (hunk 12 above)
Comment (azizbekxm) [reply docs/api-guide.md]: Fixed

**Pull request merged**